src/java.base/share/classes/java/lang/invoke/VarHandle.java

```diff
@@ -2198,15 +2206,32 @@ public Optional<VarHandleDesc> describeConstable() {
     @Stable
     MethodHandle[] methodHandleTable;
 
+    private static final long METHOD_HANDLE_TABLE_OFFSET;
+
+    static {
+        METHOD_HANDLE_TABLE_OFFSET = UNSAFE.objectFieldOffset(VarHandle.class, "methodHandleTable");
+    }
+
     @ForceInline
     final MethodHandle getMethodHandle(int mode) {
         MethodHandle[] mhTable = methodHandleTable;
         if (mhTable == null) {
-            mhTable = methodHandleTable = new MethodHandle[AccessMode.COUNT];
+            mhTable = new MethodHandle[AccessMode.COUNT];
+            if (!UNSAFE.compareAndSetReference(this, METHOD_HANDLE_TABLE_OFFSET, null, mhTable)) {
+                // We lost the race. Use the winning table instead.
+                mhTable = methodHandleTable;
+                assert mhTable != null;
+            }
```

Comment:
Suggest this should use `compareAndExchangeReference()` which avoids an unnecessary access and makes it clear the result is non-null:
```java
            Object other = UNSAFE.compareAndExchangeReference(this, METHOD_HANDLE_TABLE_OFFSET, null, mhTable);
            if (other != null) {
                // We lost the race. Use the winning table instead.
                mhTable = (MethodHandle[])other;
            }
```
Likewise for element access below:
```java
            long offset = Unsafe.ARRAY_OBJECT_BASE_OFFSET + (Unsafe.ARRAY_OBJECT_INDEX_SCALE * mode);
            Object other = UNSAFE.compareAndExchangeReference(mhTable, offset, null, mh);
            if (other != null) {
                // We lost the race. Use the winning table instead.
                mh = (MethodHandle)other;
            }
```